[Merged by Bors] - feat(RingTheory/LaurentSeries): prove that Laurent Series over a field are complete

faenuccio opened the pull request:
In this PR we provide a proof that the Laurent series over a field are complete.

Co-authored-by: María Inés de Frutos-Fernández @mariainesdff 

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

faenuccio commented:
@PatrickMassot I have closed #12179 and ported the modifications _ibid_ in this PR, where they are used to prove the completeness of Laurent series.

jcommelin commented on Mathlib/Topology/UniformSpace/Basic.lean:
```diff
@@ -136,6 +136,31 @@ theorem mem_idRel {a b : α} : (a, b) ∈ @idRel α ↔ a = b :=
 theorem idRel_subset {s : Set (α × α)} : idRel ⊆ s ↔ ∀ a, (a, a) ∈ s := by
   simp [subset_def]
 
+theorem prod_subset_idRel_Eq_singleton_left {X : Type _} {S T : Set X} (hS : S.Nonempty)
+    (hT : T.Nonempty) (h_diag : S ×ˢ T ⊆ idRel) : ∃ x, S = {x} := by
```
```suggestion
theorem eq_singleton_left_of_prod_subset_idRel {X : Type _} {S T : Set X} (hS : S.Nonempty)
    (hT : T.Nonempty) (h_diag : S ×ˢ T ⊆ idRel) : ∃ x, S = {x} := by
```

jcommelin commented on Mathlib/Topology/UniformSpace/Basic.lean:
```diff
@@ -136,6 +136,31 @@ theorem mem_idRel {a b : α} : (a, b) ∈ @idRel α ↔ a = b :=
+  replace h_diag := fun x hx y hy => (h_diag y hy x hx).symm
+  exact prod_subset_idRel_Eq_singleton_left hT hS (prod_subset_iff.mpr h_diag)
+
+theorem prod_subset_idRel_Eq_singleton {X : Type _} {S T : Set X} (hS : S.Nonempty)
```
```suggestion
theorem eq_singleton_prod_subset_idRel {X : Type _} {S T : Set X} (hS : S.Nonempty)
```

faenuccio commented on Mathlib/Topology/UniformSpace/Cauchy.lean:
```diff
@@ -787,4 +787,29 @@ theorem secondCountable_of_separable [SeparableSpace α] : SecondCountableTopolo
+    (hX : uX = ⊥) {α : Filter X} (hα : Cauchy α) : X :=
+  (DiscreteUnif.cauchy_le_pure hX hα).choose
+
+theorem DiscreteUnif.cauchy_const_eq {X : Type _} {uX : UniformSpace X} (hX : uX = ⊥)
```
Done, but I added an "of".

AntoineChambert-Loir commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -26,6 +27,11 @@ import Mathlib.RingTheory.Localization.FractionRing
 * Embedding of rational functions into Laurent series, provided as a coercion, utilizing
 the underlying `RatFunc.coeAlgHom`.
 * Study of the `X`-Adic valuation on the ring of Laurent series over a field
+* In `uniformContinuous_coeff` we show that sending a Laurent series to its `d`th coefficient in
```
Can you qualify the lemmas in the docstring? For example, `LaurentSeries.uniformContinuous_coeff`…

AntoineChambert-Loir commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+
+/- Sending a Laurent series to its `d`-th coefficient is uniformly continuous when the coefficient
+field has the discrete uniformity. -/
+theorem uniformContinuous_coeff {uK : UniformSpace K} (h : uK = ⊥) (d : ℤ) :
```
I have the (mathematical) impression that the theorem does not really need the hypothesis that the coefficient has the discrete uniformity.

AntoineChambert-Loir commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+  apply bot_uniformity ▸ h ▸ hS; rfl
+
+/-- Since extracting coefficients is uniformly continuous, every Cauchy filter in
+`laurent_series K` gives rise to a Cauchy filter in `K` for every `d : ℤ`, and such Cauchy filter
```
`LaurentSeries K`

AntoineChambert-Loir commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+vanishes for all `n ≤ N` and almost all series in the filter. This is an auxiliary lemma used
+to construct the limit of the Cauchy filter as a Laurent series, ensuring that the support of the
+limit is `PWO`.-/
+lemma Cauchy.exists_lb_eventual_support {ℱ : Filter (LaurentSeries K)} (hℱ : Cauchy ℱ) :
```
Here, I'm certain that the fact that one uses Laurent series is not needed, the same proof would probably show that Hahn series are complete.

faenuccio replied:
Well, my problem here is that I make strong use of an explicit description of the uniformity on `LaurentSeries K` coming from its `Valued` group structure, where entourages are defined in terms of the `X`-adic valuation. Nothing of this is currently available for Hahn series. I can of course make a remark (or you can add it to this PR, please do feel free!) but I fear that generalizing the result now would result in a gigantic refactor.

AntoineChambert-Loir replied:
Then a remark will be OK. Ultimately people will prove that the natural topology on Hahn series is complete, and that for Z, it gives the topology given by the valuation.

AntoineChambert-Loir commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+    (Set.inter_subset_right (s := S)) |>.trans H <| Set.mem_prod.mpr ⟨hf, hg⟩
+  exact hN g (le_of_lt h_prod)
+
+/- The support of `Cauchy.coeff` is bounded below -/
```
maybe: “has a lower bound” (to explain the discrepancy with the next theorem)

AntoineChambert-Loir commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+  by_contra hNd
+  exact hd ((exists_lb_support hℱ).choose_spec d (not_le.mp hNd))
+
+/-- To any Cauchy filter ℱ of `laurent_series K`, we can attach a laurent series that is the limit
```
`LaurentSeries K`.

AntoineChambert-Loir commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+  HahnSeries.mk (coeff hℱ) <| Set.IsWF.isPWO (coeff_support_bddBelow _).wellFoundedOn_lt
+
+/- The following lemma shows that for every `d` smaller than the minimum between the integers
+produced in `cauchy.exists_lb_eventual_support` and `cauchy.exists_lb_support`, for almost all
```
`Cauchy`. (And maybe adjust the rest)

AntoineChambert-Loir commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+  let N := min (exists_lb_coeff_ne hℱ).choose D
+  -- The goal becomes to show that the intersection of all `φ n` (for `n ≤ D`) is in `ℱ`.
+  suffices (⋂ n ∈ Set.Iio D, φ n) ∈ ℱ by
+    exact ℱ.3 this intersec
```
I would prefer to read `ℱ.sets_of_superset` rather than `ℱ.3`

AntoineChambert-Loir commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+    intro _ hm _ hd
+    exact hd _ (lt_trans hm H)
+  -- We are left with the case when some coefficients below `D` are still non-zero.
+  · rw [← Set.Iio_union_Ico_eq_Iio (le_of_not_gt H), Set.biInter_union]
```
I have the impression that the use of `H` is not that necessary (so that the `by_cases H` earlier maybe too complicated than needed).

AntoineChambert-Loir commented on Mathlib/Topology/UniformSpace/Cauchy.lean:
```diff
@@ -787,4 +787,29 @@ theorem secondCountable_of_separable [SeparableSpace α] : SecondCountableTopolo
+
+open Filter
+
+/-- A Cauchy filter in a discrete uniform space is contained in a principal filter-/
```
Here, it should suffice that the *topology* is discrete (whatever the uniform structure is).

AntoineChambert-Loir replied:
(And even a bit less : it suffices that all points be open)

AntoineChambert-Loir commented on Mathlib/Topology/UniformSpace/Cauchy.lean:
```diff
@@ -787,4 +787,29 @@ theorem secondCountable_of_separable [SeparableSpace α] : SecondCountableTopolo
+    (hX : uX = ⊥) {α : Filter X} (hα : Cauchy α) : X :=
+  (DiscreteUnif.cauchy_le_pure hX hα).choose
+
+theorem DiscreteUnif.eq_const_of_cauchy {X : Type _} {uX : UniformSpace X} (hX : uX = ⊥)
```
This should hold for the discrete topology.

AntoineChambert-Loir replied:
(NB. Points open + T2 means that points are open and closed, so the topology is discrete.)

AntoineChambert-Loir reviewed (approved):
My comments are minor, it may not be necessary that they are implemented (if they work as planned), but what holds mathematically should be recorded in the file so as to prepare for further work of other people.

faenuccio commented:
@AntoineChambert-Loir I have addressed all your points concerning the file `LaurentSeries.lean` (with a comment that generalizing to arbitrary Hahn series seems hard at the moment). I will tackle the other comments about the discrete topology tomorrow.

faenuccio commented:
I have added a remark concerning the potential extension to Hahn Series, while I think that the comment about the result holding already for the discrete topology is false as discussed with @AntoineChambert-Loir on Zulip.

AntoineChambert-Loir commented:
It looks good to me. Thanks!

faenuccio commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+of the filter. Its `d`-th coefficient is defined as the limit of `Cauchy.coeff hℱ d`, which is
+again Cauchy but valued in the discrete space `K`. That sufficiently negative coefficients vanish
+follows from `Cauchy.coeff_support_bddBelow` -/
+def Cauchy.mk_LaurentSeries {ℱ : Filter (LaurentSeries K)} (hℱ : Cauchy ℱ) : LaurentSeries K :=
```
Yes, given we're in the LaurentSeries namespace this makes sense. Updating.

faenuccio commented on Mathlib/RingTheory/LaurentSeries.lean:
```diff
@@ -609,3 +617,169 @@ theorem val_le_one_iff_eq_coe (f : LaurentSeries K) : Valued.v f ≤ (1 : ℤₘ
+open scoped Topology
+
+/- The main result showing that the Cauchy filter tends to the `hℱ.mk_LaurentSeries`-/
+theorem Cauchy.eventually_mem_nhds {ℱ : Filter (LaurentSeries K)} (hℱ : Cauchy ℱ)
```
Done, I have called `instLaurentSeriesComplete` the instance below, and mentioned it in the namespace.

AntoineChambert-Loir replied:
Do you really need to give the instance a name? Can't it be generated automatically?

faenuccio replied:
Yes, it could (and it was), but one of @jcommelin 's request was that I mention it explicitly in the docstring, and for this it needs to be named.

AntoineChambert-Loir replied:
OK. My impression was that instances get named automatically, and hence you could use/document the name even if you let Lean compute it for you. (I admit this makes the docstring sensitive to a future change in the naming process.)

faenuccio replied:
Yes, they are but their name don't appear. You can use `#synth` to see what has been created automatically, but since the name does not appear in print any textual search would fail (for now it is the last bit of the file, but if later on more things are added, it might be useful to be able to search for its name and to jump to the right line).

jcommelin commented:
Thanks :tada: 

If CI passes, please remove the label `awaiting-CI` and merge this yourself, by adding a comment `bors r+`.

bors d+

faenuccio commented:
bors r+

sgouezel commented:
bors r-

sgouezel commented:
Can you merge master, fix the build, and then send it back to bors?
bors d+

faenuccio commented:
> Can you merge master, fix the build, and then send it back to bors? bors d+

Oh sure, when I sent it to bors everything was green (or at least I believe): I wonder if it is possible that some conflict emerged while bors was merging another batch of PR's, or if I was just too sloppy, Johan's warning notwhistanding.

faenuccio commented:
bors r+
